src/views/GoTV/StreamManager.ts

```diff
@@ -0,0 +1,104 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { get } from "requests";
+import { Stream } from "GoTV";
+import * as preferences from "preferences";
+import { EventEmitter } from "events";
+import { socket } from "sockets";
+
+class StreamManager extends EventEmitter {
+    private static instance: StreamManager;
+    private streams: Stream[] = [];
+    private allowMatureStreams: boolean;
+    private selectedLanguages: string[];
+
+    private constructor() {
+        super();
+        this.allowMatureStreams = preferences.get("gotv.allow-mature-streams");
+        this.selectedLanguages = preferences.get("gotv.selected-languages");
+        this.fetchStreams();
+        preferences.watch("gotv.allow-mature-streams", this.updatePreferences.bind(this));
+        preferences.watch("gotv.selected-languages", this.updatePreferences.bind(this));
+        this.setupWebSocket();
+    }
+
+    public static getInstance(): StreamManager {
+        if (!StreamManager.instance) {
+            StreamManager.instance = new StreamManager();
+        }
+        return StreamManager.instance;
+    }
+
+    private updatePreferences() {
+        this.allowMatureStreams = preferences.get("gotv.allow-mature-streams");
+        this.selectedLanguages = preferences.get("gotv.selected-languages");
+        this.fetchStreams();
+    }
+
+    private filterStreams(streams: Stream[]): Stream[] {
+        let filteredStreams = streams;
+        if (!this.allowMatureStreams) {
+            filteredStreams = filteredStreams.filter((stream) => !stream.is_mature);
+        }
+        if (this.selectedLanguages.length > 0 && this.selectedLanguages[0] !== "") {
+            filteredStreams = filteredStreams.filter((stream) =>
+                this.selectedLanguages.includes(stream.language),
+            );
+        }
+        return filteredStreams;
+    }
+
+    public fetchStreams() {
+        get("gotv/streams")
+            .then((streams: Stream[]) => {
+                this.streams = this.filterStreams(streams);
+                this.emit("update", this.streams);
+            })
+            .catch((error) => {
+                console.error("Error fetching streams:", error);
+            });
+    }
+
+    private setupWebSocket() {
+        socket.on("ui-push", (msg) => {
+            if (msg.event === "update_streams") {
+                const updatedStreams = JSON.parse(msg.data);
+                this.streams = this.filterStreams(updatedStreams);
+                this.emit("update", this.streams);
+            }
+        });
+
+        socket.on("connect", () => {
+            socket.send("ui-pushes/subscribe", { channel: "gotv" });
+        });
+    }
+
+    public getStreams(): Stream[] {
+        return this.streams;
+    }
+
+    public subscribe(callback: (streams: Stream[]) => void) {
```

Comment:
Elsewhere in the code we just let folks use the event emitter interface directly, so for consistency let's get rid of the `subscribe` / `unsubscribe` functions and have folks just do `streamManager.on("update", (streams) => {...})`